# Wragle Twitter data for We Rate Dogs twitter account
Going to use tweepy package to do this. The below code initializes 

In [1]:
import tweepy

# import twitter keys and tokens required for initializing api
from twitter_api_keys import consumer_key, consumer_secret, access_token, access_secret

# the above line will not work for you if you just clone the repo.
# To make it work, set the above variables to your specific values that you get when you
# create your twitter app.


auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)

## Gather
Import necessary libraries and read the json twitter data, as well as the 'twitter-archive-enhanced.csv'.
This csv was provided as part of the project and was downloaded manually from the link: https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59a4e958_twitter-archive-enhanced/twitter-archive-enhanced.csv

In [2]:
import pandas as pd

In [3]:
tweet_archive_df = pd.read_csv('twitter-archive-enhanced.csv')
tweet_archive_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [4]:
tweet_archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

#### Checking the json data for the first tweet id just to check


In [5]:
check_tweet = api.get_status(tweet_archive_df.tweet_id[0], tweet_mode='extended')

In [6]:
# Checking all available methods for the status object
dir(check_tweet)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_api',
 '_json',
 'author',
 'contributors',
 'coordinates',
 'created_at',
 'destroy',
 'display_text_range',
 'entities',
 'extended_entities',
 'favorite',
 'favorite_count',
 'favorited',
 'full_text',
 'geo',
 'id',
 'id_str',
 'in_reply_to_screen_name',
 'in_reply_to_status_id',
 'in_reply_to_status_id_str',
 'in_reply_to_user_id',
 'in_reply_to_user_id_str',
 'is_quote_status',
 'lang',
 'parse',
 'parse_list',
 'place',
 'possibly_sensitive',
 'possibly_sensitive_appealable',
 'retweet',
 'retweet_count',
 'retweeted',
 'retweets',
 'source',
 'source_url',
 'truncated',
 'user']

In [7]:
# using the full_text attribute of the status object
check_tweet.full_text

"This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU"

In [8]:
import json

In [9]:
type(check_tweet)

tweepy.models.Status

In [10]:
str(check_tweet)

'Status(_api=<tweepy.api.API object at 0x000001CF79D3F630>, _json={\'created_at\': \'Tue Aug 01 16:23:56 +0000 2017\', \'id\': 892420643555336193, \'id_str\': \'892420643555336193\', \'full_text\': "This is Phineas. He\'s a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU", \'truncated\': False, \'display_text_range\': [0, 85], \'entities\': {\'hashtags\': [], \'symbols\': [], \'user_mentions\': [], \'urls\': [], \'media\': [{\'id\': 892420639486877696, \'id_str\': \'892420639486877696\', \'indices\': [86, 109], \'media_url\': \'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg\', \'media_url_https\': \'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg\', \'url\': \'https://t.co/MgUWQ76dJU\', \'display_url\': \'pic.twitter.com/MgUWQ76dJU\', \'expanded_url\': \'https://twitter.com/dog_rates/status/892420643555336193/photo/1\', \'type\': \'photo\', \'sizes\': {\'thumb\': {\'w\': 150, \'h\': 150, \'resize\': \'crop\'}, \'medium\': {\'w\': 540, \'h\': 528, \

In [11]:
tweet_archive_df.tweet_id.head().values

array([892420643555336193, 892177421306343426, 891815181378084864,
       891689557279858688, 891327558926688256], dtype=int64)

In [12]:
json_str_list = []
exception_list = []
import sys

## Testing the code to download the twitter json data on sample

```
with open('tweet-json-test.txt','w') as json_file:
    for tweetid in tweet_archive_df.tweet_id.tail().values:
        print('getting data for: ',tweetid)
        try:
            status = api.get_status(tweetid, tweet_mode='extended')
            json_str_list.append( json.dumps(status._json) )
        except:
            exception_list.append(tweetid)
            print('exception_raised for: ',tweetid)
            print(sys.exc_info()[0])
    json.dump(json_str_list,json_file)

 ```

### Test

```
with open('tweet-json-test.txt','r') as json_file:
    data = json.load(json_file)
print(len(data))

#json.load
```

## Gathering all twitter data for the given tweet ids
The above programmatic download code for the twitter json data will now be used to download all 2356 tweet ids.
<br> The tweet ids that have been deleted will be catched in the exception list, which I will check later

In [13]:
len(tweet_archive_df)

2356

```python
with open('tweet-json.txt','w') as json_file:
    for tweetid in tweet_archive_df.tweet_id.values:
        print('getting data for: ',tweetid)
        try:
            status = api.get_status(tweetid, tweet_mode='extended')
            json_str_list.append( json.dumps(status._json) )
        except:
            exception_list.append(tweetid)
            print('exception_raised for: ',tweetid)
            print(sys.exc_info()[0])
    json.dump(json_str_list,json_file)

```

### Test
Reading the stored json data into pandas dataframe

len(exception_list)

21 twitter ids from the archive do not have existing data from

In [14]:
status = []
with open('tweet-json.txt','r') as file:
    for line in file:
        status.append(json.loads(line))


In [15]:
dicts = []
for string in status[0]:
    dicts.append(json.loads(string))

In [16]:
type(dicts[0])

dict

In [17]:
tweet_df = pd.DataFrame(dicts)
tweet_df.head()

contributors coordinates                      created_at display_text_range  \
0         None        None  Mon Nov 16 00:24:50 +0000 2015           [0, 120]   
1         None        None  Mon Nov 16 00:04:52 +0000 2015           [0, 137]   
2         None        None  Sun Nov 15 23:21:54 +0000 2015           [0, 130]   
3         None        None  Sun Nov 15 23:05:30 +0000 2015           [0, 139]   
4         None        None  Sun Nov 15 22:32:08 +0000 2015           [0, 131]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 666049244999131136, 'id_str'...             104   
1  {'media': [{'id': 666044217047650304, 'id_str'...             288   
2  {'media': [{'id': 666033409081393153, 'id_str'...             120   
3  {'media': [{'id': 666029276303482880, 'id_str'...             124   
4  {'media': [{'id': 666020881337073664, 'id_str'...            2503   

   favorited                                          full_text   geo  \
0      False  Here we have a 1949 1st generation vulpix. Enj...  None   
1      False  This is a purebred Piers Morgan. Loves to Netf...  None   
2      False  Here is a very happy pup. Big fan of well-main...  None   
3      False  This is a western brown Mitsubishi terrier. Up...  None   
4      False  Here we have a Japanese Irish Setter. Lost eye...  None   

                         ...                          quoted_status  \
0                        ...                                    NaN   
1                        ...                                    NaN   
2                        ...                                    NaN   
3                        ...                                    NaN   
4                        ...                                    NaN   

  quoted_status_id quoted_status_id_str  quoted_status_permalink  \
0              NaN                  NaN                      NaN   
1              NaN                  NaN                      NaN   
2              NaN                  NaN                      NaN   
3              NaN                  NaN                      NaN   
4              NaN                  NaN                      NaN   

  retweet_count  retweeted retweeted_status  \
0            41      False              NaN   
1           135      False              NaN   
2            43      False              NaN   
3            46      False              NaN   
4           493      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

## Gather the image predictions dataset using requests
download the dataset from https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv

In [18]:
import requests

In [19]:
response = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')

In [20]:
response

<Response [200]>

In [21]:
with open('image-predictions.tsv','w') as file:
    file.write(response.text)

In [22]:
image_pred_df = pd.read_csv('image-predictions.tsv',sep = '\t')
image_pred_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

## Gathered Data
- tweet_df -contains tweet data from querying each tweet id from the archive dataset
- tweet_archive_df -contains all the tweets of we rate dogs, that contain some rating for a dog
- imge_pred_df -contains a dataframe of predictions of the breed of each dog in the tweet archive dataframe

# Assess

In [23]:
tweet_df.head()

contributors coordinates                      created_at display_text_range  \
0         None        None  Mon Nov 16 00:24:50 +0000 2015           [0, 120]   
1         None        None  Mon Nov 16 00:04:52 +0000 2015           [0, 137]   
2         None        None  Sun Nov 15 23:21:54 +0000 2015           [0, 130]   
3         None        None  Sun Nov 15 23:05:30 +0000 2015           [0, 139]   
4         None        None  Sun Nov 15 22:32:08 +0000 2015           [0, 131]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 666049244999131136, 'id_str'...             104   
1  {'media': [{'id': 666044217047650304, 'id_str'...             288   
2  {'media': [{'id': 666033409081393153, 'id_str'...             120   
3  {'media': [{'id': 666029276303482880, 'id_str'...             124   
4  {'media': [{'id': 666020881337073664, 'id_str'...            2503   

   favorited                                          full_text   geo  \
0      False  Here we have a 1949 1st generation vulpix. Enj...  None   
1      False  This is a purebred Piers Morgan. Loves to Netf...  None   
2      False  Here is a very happy pup. Big fan of well-main...  None   
3      False  This is a western brown Mitsubishi terrier. Up...  None   
4      False  Here we have a Japanese Irish Setter. Lost eye...  None   

                         ...                          quoted_status  \
0                        ...                                    NaN   
1                        ...                                    NaN   
2                        ...                                    NaN   
3                        ...                                    NaN   
4                        ...                                    NaN   

  quoted_status_id quoted_status_id_str  quoted_status_permalink  \
0              NaN                  NaN                      NaN   
1              NaN                  NaN                      NaN   
2              NaN                  NaN                      NaN   
3              NaN                  NaN                      NaN   
4              NaN                  NaN                      NaN   

  retweet_count  retweeted retweeted_status  \
0            41      False              NaN   
1           135      False              NaN   
2            43      False              NaN   
3            46      False              NaN   
4           493      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

In [24]:
tweet_df.columns

Index(['contributors', 'coordinates', 'created_at', 'display_text_range',
       'entities', 'extended_entities', 'favorite_count', 'favorited',
       'full_text', 'geo', 'id', 'id_str', 'in_reply_to_screen_name',
       'in_reply_to_status_id', 'in_reply_to_status_id_str',
       'in_reply_to_user_id', 'in_reply_to_user_id_str', 'is_quote_status',
       'lang', 'place', 'possibly_sensitive', 'possibly_sensitive_appealable',
       'quoted_status', 'quoted_status_id', 'quoted_status_id_str',
       'quoted_status_permalink', 'retweet_count', 'retweeted',
       'retweeted_status', 'source', 'truncated', 'user'],
      dtype='object')

In [25]:
tweet_archive_df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [26]:
tweet_archive_df[['doggo','floofer','pupper','puppo']].describe()

doggo floofer pupper puppo
count   2356    2356   2356  2356
unique     2       2      2     2
top     None    None   None  None
freq    2259    2346   2099  2326

In [27]:
tweet_archive_df.doggo.value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [28]:
tweet_archive_df.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [29]:
tweet_archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [30]:
tweet_archive_df.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [31]:
tweet_archive_df.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [32]:
tweet_archive_df['text'].value_counts()

This is Gilbert. He's being chased by a battalion of miniature floof cows. 10/10 we all believe in you Gilbert https://t.co/wayKZkDRTG                               1
This is Berb. He just found out that they have made 31 Kidz Bop CD's. Downright terrifying. 7/10 hang in there Berb https://t.co/CIFLjiTFwZ                          1
This is Bauer. He had nothing to do with the cookies that disappeared. 13/10 very good boy https://t.co/AIMF8ouzvl                                                   1
Say hello to Alice. I'm told she enjoys car rides and smells good. 12/10 would give her everything she could ever want https://t.co/yT4vw8y77x                       1
This is Riley. She's just an adorable football fan. 12/10 I'd watch the sports with her https://t.co/kLV8zUCfc8                                                      1
This is Sebastian. He's super h*ckin fluffy. That's really all you need to know. 11/10 would snug intensely https://t.co/lqr0NdtwQo                                  

In [33]:
image_pred_df

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [34]:
image_pred_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


## Issues
### Tidyness
1. retweet and favorite count of tweet_df not joined with tweet_archive_df

##### tweet_archive_df
1. stage of dog is 1 column but is divided into 4 columns: doggo, floofer, pupper, puppo

### Quality
##### tweet_archive_df
1. several dogs do not have any dog stage assigned.
2. Dog stage Floof is named floofer
3. timestamp column is of object type and not datetime
4. missing values in dataset are replaced by 'None' text. e.g.doggo, floofer, pupper, puppo
5. retweeted status timestamp is of object type and not datetime
6. rating_denominator values have values other than 10
7. rating_numerator values have some values greater than 100
8. There are 181 retweeted tweets, and not original ratings. found from retweeted_to column
9. Some of the rating numerators are wrongly extracted from text. e.g.row:314, numerator is extracted as 960, but actually 13
10. Some tweets have more than 1 dog stage

##### image_pred_df
1. There are 2075 tweets that contain images.

# Clean
## tidyness issue 1: merging datasets
### Define
Merge the datasets using the pandas merge method. use inner join, so only those rows that are common to both the tables remain.

<br> Since only 21 rows do not have corresponding data in the tweet_df out of a total of 2356 entries, this is not a significant omission.
### Code

##### Merging tweet datasets
Bringing retweet and favorite counts into the tweet archive dataframe

In [35]:
tweet_archive_clean_df = tweet_archive_df.copy()


In [36]:
tweet_archive_clean_df = tweet_archive_clean_df.merge(tweet_df[['id','retweet_count','favorite_count']],how='inner',left_on='tweet_id',right_on='id',
              right_index=False)

### Test

In [37]:
tweet_archive_clean_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  \
0                  10   Phineas  None    None   None  None   
1                  10     Tilly  None    None   None  None   
2                  10    Archie  None    None   None  None   
3                  10     Darla  None    None   None  None   
4                  10  Franklin  None    None   None  None   

                   id  retweet_count  favorite_count  
0  892420643555336193           8178           37517  
1  892177421306343426           6051           32258  
2  891815181378084864           4003           24298  
3  891689557279858688           8318           40849  
4  891327558926688256           9025           39062

In [38]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2345 entries, 0 to 2344
Data columns (total 20 columns):
tweet_id                      2345 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2345 non-null object
source                        2345 non-null object
text                          2345 non-null object
retweeted_status_id           166 non-null float64
retweeted_status_user_id      166 non-null float64
retweeted_status_timestamp    166 non-null object
expanded_urls                 2286 non-null object
rating_numerator              2345 non-null int64
rating_denominator            2345 non-null int64
name                          2345 non-null object
doggo                         2345 non-null object
floofer                       2345 non-null object
pupper                        2345 non-null object
puppo                         2345 non-null object
id                            23

## Tidyness issue 2: Stage of dog is distributed into 4 columns
## Quality issue 4: Handle 'None' values in dog stage columns

### Define
- combine the 4 columns, keep only non none values

### Code

In [39]:
c = tweet_archive_clean_df[['doggo','floofer','pupper','puppo']].replace('None','')
c['stage'] = c.doggo + c.floofer + c.pupper + c.puppo
c.stage.value_counts()

                1969
pupper           242
doggo             82
puppo             29
doggopupper       12
floofer            9
doggopuppo         1
doggofloofer       1
Name: stage, dtype: int64

In [40]:
tweet_archive_clean_df[c.stage == 'floofer'].text.values

array(['Meet Grizzwald. He may be the floofiest floofer I ever did see. Lost eyes saving a schoolbus from a volcano erpuption. 13/10 heroic as h*ck https://t.co/rf661IFEYP',
       'This is Doc. He takes time out of every day to worship our plant overlords. 12/10 quite the floofer https://t.co/azMneS6Ly5',
       "Atlas rolled around in some chalk and now he's a magical rainbow floofer. 13/10 please never take a bath https://t.co/nzqTNw0744",
       "This is Blu. He's a wild bush Floofer. I wish anything made me as happy as bushes make Blu. 12/10 would frolic with https://t.co/HHUAnBb6QB",
       "Here's a golden floofer helping with the groceries. Bed got in way. Still 11/10 helpful af (vid by @categoen) https://t.co/6ZRoZUWFmd",
       "Just wanted to share this super rare Rainbow Floofer in case you guys haven't seen it yet. 13/10 colorful af https://t.co/CaG9MzD3WT",
       "This is Moose. He's a Polynesian Floofer. Dapper af. 10/10 would pet diligently https://t.co/mVfqRdppTL",
  

In [41]:
tweet_archive_clean_df['stage'] = c.stage

### Test

In [42]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2345 entries, 0 to 2344
Data columns (total 21 columns):
tweet_id                      2345 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2345 non-null object
source                        2345 non-null object
text                          2345 non-null object
retweeted_status_id           166 non-null float64
retweeted_status_user_id      166 non-null float64
retweeted_status_timestamp    166 non-null object
expanded_urls                 2286 non-null object
rating_numerator              2345 non-null int64
rating_denominator            2345 non-null int64
name                          2345 non-null object
doggo                         2345 non-null object
floofer                       2345 non-null object
pupper                        2345 non-null object
puppo                         2345 non-null object
id                            23

## Quality issue 1: Handle missing dog stage data
### Define
- replace the missing values with 'not_mentioned'

### Code

In [43]:
tweet_archive_clean_df.stage.replace('','not_mentioned',inplace=True)

### Test

In [44]:
tweet_archive_clean_df.stage.value_counts()

not_mentioned    1969
pupper            242
doggo              82
puppo              29
doggopupper        12
floofer             9
doggopuppo          1
doggofloofer        1
Name: stage, dtype: int64

## Quality issue 2: dog stage floof is mentioned as floofer
### Define
- replace every occurance of floofer with floof in dog stage column

### Code

In [45]:
tweet_archive_clean_df.stage.replace('floofer','floof',inplace=True)

### Test

In [46]:
tweet_archive_clean_df.stage.value_counts()

not_mentioned    1969
pupper            242
doggo              82
puppo              29
doggopupper        12
floof               9
doggopuppo          1
doggofloofer        1
Name: stage, dtype: int64

## Quality issue 3: Time stamp is of type object
### Define
- Convert time stamp column to date format

### Code

In [47]:
tweet_archive_clean_df.timestamp = pd.to_datetime(tweet_archive_clean_df.timestamp)

### Test

In [48]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2345 entries, 0 to 2344
Data columns (total 21 columns):
tweet_id                      2345 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2345 non-null datetime64[ns]
source                        2345 non-null object
text                          2345 non-null object
retweeted_status_id           166 non-null float64
retweeted_status_user_id      166 non-null float64
retweeted_status_timestamp    166 non-null object
expanded_urls                 2286 non-null object
rating_numerator              2345 non-null int64
rating_denominator            2345 non-null int64
name                          2345 non-null object
doggo                         2345 non-null object
floofer                       2345 non-null object
pupper                        2345 non-null object
puppo                         2345 non-null object
id                      

## Issues left


### Quality
##### tweet_archive_df
5. retweeted status timestamp is of object type and not datetime
6. rating_denominator values have values other than 10
7. rating_numerator values have some values greater than 100
8. There are 181 retweeted tweets, and not original ratings. found from retweeted_to column
9. Some of the rating numerators are wrongly extracted from text. e.g.row:314, numerator is extracted as 960, but actually 13
10. Some tweets have more than 1 dog stage

##### image_pred_df
1. There are 2075 tweets that contain images.

## Quality issue 5: retweeted Time stamp is of type object
### Define
- Convert retweeted time stamp column to date format

### Code

In [49]:
tweet_archive_clean_df.retweeted_status_timestamp = pd.to_datetime(tweet_archive_clean_df.retweeted_status_timestamp)

### Test

In [50]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2345 entries, 0 to 2344
Data columns (total 21 columns):
tweet_id                      2345 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2345 non-null datetime64[ns]
source                        2345 non-null object
text                          2345 non-null object
retweeted_status_id           166 non-null float64
retweeted_status_user_id      166 non-null float64
retweeted_status_timestamp    166 non-null datetime64[ns]
expanded_urls                 2286 non-null object
rating_numerator              2345 non-null int64
rating_denominator            2345 non-null int64
name                          2345 non-null object
doggo                         2345 non-null object
floofer                       2345 non-null object
pupper                        2345 non-null object
puppo                         2345 non-null object
id              

## Quality issue 6: denominator values of rating have varied values other than 10
### Define
- check the genuineness of the ratings
- extract the correct denominator and replace in the necessary column

### Code

In [51]:
c = tweet_archive_clean_df.copy()
#c.rating_denominator.value_counts()
c[c.rating_denominator!=10].text.values[0]

"@jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho"

In [52]:
c[c.rating_denominator!=10].text.values[1]

'@docmisterio account started on 11/15/15'

In [53]:
c[c.rating_denominator!=10].text.values[2]

'The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd'

In [54]:
c[c.rating_denominator!=10].text.values[3]

'Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx'

## Iterate on Assess Step
As we see there are some corrections that are required in the denominator of the rating, we are iterating on our assess step once again.

## Assessing rating_denominator

In [55]:
c[c.rating_denominator >10].rating_denominator.value_counts()

50     3
20     2
80     2
11     2
90     1
120    1
150    1
16     1
15     1
110    1
170    1
40     1
70     1
130    1
Name: rating_denominator, dtype: int64

In [56]:
print(c[c.rating_denominator >10].text.values)

['@docmisterio account started on 11/15/15'
 'The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd'
 'Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE'
 'After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ'
 'Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv'
 'Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a'
 'This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq'
 "Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1"
 "Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12"
 'From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 w

In [57]:
print(c[c.rating_denominator ==11].text.values)
# Both these are not actually 11, but are 10

['After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ'
 'This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5']


In [58]:
print(c[c.rating_denominator ==50].text.values)
# Both these are not actually 11, but are 10

['This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq'
 'From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK'
 "Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa"]


In [59]:
print(c[c.rating_denominator <10].text.values)

["@jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho"
 'Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx'
 'This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv']


Assessing from the above, we see that ratings can have more or less than 10 as denominator

In [60]:
c[c.rating_denominator ==0].text.values

array(["@jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho"],
      dtype=object)

In [61]:
c[c.rating_denominator ==2].text.values

array(['This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv'],
      dtype=object)

## More issues
#### tweet_archive_df
- rating_denominator is actually 10 where it shows 0
- rating_denominator is actually 10 where it shows 2
- rating_denominator is actually 10 where it shows 11 (both the cases)
- rating_denominator for Bluebert is not 50, it is 10

## Clean
### Issue: rating numerator has values greater than 100, and rating denominator has values other than 10
This issues have been checked to be non issues, as there is no fixed range for ratings to exist

## Issues left


### Quality
##### tweet_archive_df
8. There are 181 retweeted tweets, and not original ratings. found from retweeted_to column
9. Some of the rating numerators are wrongly extracted from text. e.g.row:314, numerator is extracted as 960, but actually 13
10. Some tweets have more than 1 dog stage

##### image_pred_df
1. There are 2075 tweets that contain images.

## Clean
### Issue: 181 retweeted tweets exist
#### Define
- Remove tweets that are retweets

#### Code

In [62]:
len(tweet_archive_clean_df[-tweet_archive_clean_df.retweeted_status_id.isna()])

166

In [63]:
tweet_archive_clean_df = tweet_archive_clean_df[tweet_archive_clean_df.retweeted_status_id.isna()]

#### Test

In [64]:
len(tweet_archive_clean_df[-tweet_archive_clean_df.retweeted_status_id.isna()])

0

## Quality issue: Rating numerators are wrongly extracted for some
### Define
- check the deviations (assess again)

### Code

## Assess rating numerator

In [65]:
tweet_archive_clean_df.rating_numerator.value_counts()

12      498
10      441
11      425
13      305
9       158
8       100
7        56
14       43
5        38
6        34
3        19
4        17
2         9
1         8
420       2
0         2
20        1
24        1
26        1
84        1
44        1
50        1
60        1
80        1
165       1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
15        1
17        1
27        1
45        1
75        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

since, usually 'we rate dogs' does not give a rating below 10, lets further investigate these cases

In [66]:
tweet_archive_clean_df[tweet_archive_clean_df.rating_numerator <10].text.values

array(['This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948',
       "This is Jerry. He's doing a distinguished tongue slip. Slightly patronizing tbh. You think you're better than us, Jerry? 6/10 hold me back https://t.co/DkOBbwulw1",
       "When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag",
       "I was going to do 007/10, but the joke wasn't worth the &lt;10 rating",
       'Who keeps sending in pictures without dogs in them? This needs to stop. 5/10 for the mediocre road https://t.co/ELqelxWMrC',
       "This is Wesley. He's clearly trespassing. Seems rather h*ckin violent too. Weaponized forehead. 3/10 wouldn't let in https://t.co/pL7wbMRW7M",
       'Another pic without a dog in it? What am I supposed to do? Rate the carpet? Fine I will. 7/10 looks adequately comfy https://t.co/OJZQ6I4gGd',
       "Meet Gerald. He's a fairly exotic

In [67]:
tweet_archive_clean_df[tweet_archive_clean_df.rating_numerator ==5].text.values

array(['This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948',
       'Who keeps sending in pictures without dogs in them? This needs to stop. 5/10 for the mediocre road https://t.co/ELqelxWMrC',
       "Please stop sending it pictures that don't even have a doggo or pupper in them. Churlish af. 5/10 neat couch tho https://t.co/u2c9c7qSg8",
       "This is Dave. He's a tropical pup. Short lil legs (dachshund mix?) Excels underwater, but refuses to eat kibble 5/10 https://t.co/ZJnCxlIf62",
       'Please only send in dogs. This t-rex is very scary. 5/10 ...might still pet (vid by @helizabethmicha) https://t.co/Vn6w5w8TO2',
       "When bae says they can't go out but you see them with someone else that same night. 5/10 &amp; 10/10 for heartbroken pup https://t.co/aenk0KpoWM",
       "Army of water dogs here. None of them know where they're going. Have no real purpose. Aggressive barks. 5/10 for all https:

## Quality issues
- when rating numerator is not an integer, then it shows only the last digit of the numerator

## Issue: numerators are wrong where they are in decimals
### Define
- re-extract rating numerators using regular expressions
- keep the last occurance of the numerator and denominator

### Code

In [68]:
tweet_archive_clean_df[tweet_archive_clean_df.rating_numerator==5].text.values[0]

'This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948'

In [69]:
c = tweet_archive_clean_df.text.str.extractall( r'\s([0-9./]+)/[0-9./"]+' )
d = tweet_archive_clean_df.copy()
#c.reset_index('match')[c.reset_index('match').match==0].iloc[:,1]
# c.reset_index('match',inplace=True)

# d['num_edit'] = c[c.match==0].iloc[:,1]
# #d.head()
# #d[c.match==1].num_edit = c[c.match==1].iloc[:,1]
# d.head()

In [70]:
d['num_edit'] = d.text.str.extractall( r'[\s.;]*(0|[1-9][0-9.]*)/[1-9][0-9]*[\w\W]*$' ).reset_index('match').iloc[:,1]

d['den_edit'] = d.text.str.extractall( r'[\s.;]*[0-9.]*/([1-9][0-9]*)[\w\W]*$' ).reset_index('match').iloc[:,1]

d.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2179 entries, 0 to 2344
Data columns (total 23 columns):
tweet_id                      2179 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2179 non-null datetime64[ns]
source                        2179 non-null object
text                          2179 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null datetime64[ns]
expanded_urls                 2121 non-null object
rating_numerator              2179 non-null int64
rating_denominator            2179 non-null int64
name                          2179 non-null object
doggo                         2179 non-null object
floofer                       2179 non-null object
pupper                        2179 non-null object
puppo                         2179 non-null object
id                    

In [71]:
d.num_edit.value_counts()

12       498
10       441
11       425
13       306
9        158
8        100
7         56
14        43
5         36
6         34
3         19
4         17
2          9
1          8
420        2
0          2
165        1
144        1
1776       1
15         1
84         1
13.5       1
60         1
143        1
44         1
11.27      1
204        1
20         1
99         1
17         1
88         1
24         1
9.75       1
11.26      1
45         1
666        1
80         1
9.5        1
50         1
121        1
182        1
Name: num_edit, dtype: int64

In [72]:
d.den_edit.value_counts()

10     2158
50        3
11        2
20        2
80        2
90        1
15        1
7         1
110       1
16        1
170       1
120       1
40        1
70        1
150       1
2         1
130       1
Name: den_edit, dtype: int64

In [73]:
tweet_archive_clean_df.rating_numerator = d.num_edit.astype('float')
tweet_archive_clean_df.rating_denominator = d.den_edit.astype('float')

### Test

In [74]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2179 entries, 0 to 2344
Data columns (total 21 columns):
tweet_id                      2179 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2179 non-null datetime64[ns]
source                        2179 non-null object
text                          2179 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null datetime64[ns]
expanded_urls                 2121 non-null object
rating_numerator              2179 non-null float64
rating_denominator            2179 non-null float64
name                          2179 non-null object
doggo                         2179 non-null object
floofer                       2179 non-null object
pupper                        2179 non-null object
puppo                         2179 non-null object
id                

### Issue: rating_denominator errors
#### Define
- change rating_denominator values from:
    - 0 to 10
    - 2 to 10
    - 11 to 10
    - 50 to 10 (for name Bluebert)

#### Code

In [75]:
tweet_archive_clean_df.rating_denominator.replace(0,10,inplace=True)
tweet_archive_clean_df.rating_denominator.replace(2,10,inplace=True)
tweet_archive_clean_df.rating_denominator.replace(11,10,inplace=True)

In [76]:
tweet_archive_clean_df.loc[tweet_archive_clean_df.name=='Bluebert','rating_denominator'] = 10

#### Test

In [77]:
tweet_archive_clean_df.rating_denominator.value_counts()

10.0     2162
50.0        2
80.0        2
20.0        2
150.0       1
110.0       1
90.0        1
130.0       1
70.0        1
170.0       1
120.0       1
16.0        1
40.0        1
7.0         1
15.0        1
Name: rating_denominator, dtype: int64

## Issues left


### Quality
##### tweet_archive_df
10. Some tweets have more than 1 dog stage

##### image_pred_df
1. There are 2075 tweets that contain images.

## Clean
### Issue: tweets have 2 dog stages
#### Define
- rename dog stage, separating each dog stage

#### Code

In [78]:
tweet_archive_clean_df.stage.value_counts()

not_mentioned    1837
pupper            222
doggo              75
puppo              24
doggopupper        10
floof               9
doggopuppo          1
doggofloofer        1
Name: stage, dtype: int64

In [79]:
tweet_archive_clean_df.stage.replace({ 'doggopupper':'doggo,pupper',
                               'doggopuppo':'doggo,puppo',
                               'doggofloofer':'doggo,floof'},inplace=True)

### Test

In [80]:
tweet_archive_clean_df.stage.value_counts()

not_mentioned    1837
pupper            222
doggo              75
puppo              24
doggo,pupper       10
floof               9
doggo,floof         1
doggo,puppo         1
Name: stage, dtype: int64

## Clean
### Issue: only 2075 tweets have images
#### Define
- join tweet_archive df with image predictions and keep only matching rows
>This is so that we have a final dataset of only those tweets that have ratings, are original, and have a corresponding image

#### Code

In [81]:
tweet_archive_clean_df = tweet_archive_clean_df.merge(image_pred_df[['tweet_id']], on='tweet_id')

#### Test

In [82]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1998 entries, 0 to 1997
Data columns (total 21 columns):
tweet_id                      1998 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     1998 non-null datetime64[ns]
source                        1998 non-null object
text                          1998 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null datetime64[ns]
expanded_urls                 1998 non-null object
rating_numerator              1998 non-null float64
rating_denominator            1998 non-null float64
name                          1998 non-null object
doggo                         1998 non-null object
floofer                       1998 non-null object
pupper                        1998 non-null object
puppo                         1998 non-null object
id                

## Further issues through assess

### Quality
#### tweet_archive_df
- retweeted columns are unnecessary for analysis
- doggo, floofer, pupper and puppo columns are unnecessary
- id is a duplicate column
- stage column is object and not categorical
- stage column_name is not easily understandable

## Clean
### Issue: unnecessary retweeted columns
### Define
- drop retweeted columns inplace

### Code

In [83]:
tweet_archive_clean_df.drop(['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'],axis=1,inplace=True)

### Test

In [84]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1998 entries, 0 to 1997
Data columns (total 18 columns):
tweet_id                 1998 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1998 non-null datetime64[ns]
source                   1998 non-null object
text                     1998 non-null object
expanded_urls            1998 non-null object
rating_numerator         1998 non-null float64
rating_denominator       1998 non-null float64
name                     1998 non-null object
doggo                    1998 non-null object
floofer                  1998 non-null object
pupper                   1998 non-null object
puppo                    1998 non-null object
id                       1998 non-null int64
retweet_count            1998 non-null int64
favorite_count           1998 non-null int64
stage                    1998 non-null object
dtypes: datetime64[ns](1), float64(4), int64(4), object(9)
me

### Issue: unnecessary dog stage columns
### Define
- drop doggo, puppo, pupper, floofer columns inplace

### Code

In [85]:
tweet_archive_clean_df.drop(['doggo','floofer','pupper','puppo'],axis=1,inplace=True)

### Test

In [86]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1998 entries, 0 to 1997
Data columns (total 14 columns):
tweet_id                 1998 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1998 non-null datetime64[ns]
source                   1998 non-null object
text                     1998 non-null object
expanded_urls            1998 non-null object
rating_numerator         1998 non-null float64
rating_denominator       1998 non-null float64
name                     1998 non-null object
id                       1998 non-null int64
retweet_count            1998 non-null int64
favorite_count           1998 non-null int64
stage                    1998 non-null object
dtypes: datetime64[ns](1), float64(4), int64(4), object(5)
memory usage: 234.1+ KB


### Issue: stage column name is not named correct and it is not categorical, while it should be
### Define
- rename stage column name to dog_stage
- convert dog_stage from object to categorical

### Code

In [90]:
tweet_archive_clean_df.rename({'stage':'dog_stage'},axis=1,inplace=True)

In [94]:
tweet_archive_clean_df.dog_stage = tweet_archive_clean_df.dog_stage.astype('category')

### Test

In [95]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1998 entries, 0 to 1997
Data columns (total 14 columns):
tweet_id                 1998 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1998 non-null datetime64[ns]
source                   1998 non-null object
text                     1998 non-null object
expanded_urls            1998 non-null object
rating_numerator         1998 non-null float64
rating_denominator       1998 non-null float64
name                     1998 non-null object
id                       1998 non-null int64
retweet_count            1998 non-null int64
favorite_count           1998 non-null int64
dog_stage                1998 non-null category
dtypes: category(1), datetime64[ns](1), float64(4), int64(4), object(4)
memory usage: 220.9+ KB


### Export cleaned dataset: tweet_archive_clean_df

In [98]:
tweet_archive_clean_df.to_csv('twitter_archive_master.csv',index=False)